Anlab.Mvc/Controllers/PaymentController.cs

```diff
@@ -0,0 +1,337 @@
+                return RedirectToAction("Pay", new {id = order.Id});
+            }
+
+            //Should be good,   
+            if (order.Status == OrderStatusCodes.AwaitingPayment)
+            {
+                order.Status = OrderStatusCodes.Paid;
+                _context.SaveChanges(true);
+            }
+            Message = "Payment Processed. Thank You.";
+
+            //ViewBag.PaymentDictionary = dictionary; //Debugging. Remove when not needed
+
+            return View(response);
+        }
+
+        [HttpPost]
+        [IgnoreAntiforgeryToken]
+        public ActionResult Cancel(ReceiptResponseModel response)
+        {
+            {
+                Log.ForContext("response", response, true).Information("Cancel response received");
+
+                // check signature
+                var dictionary = Request.Form.ToDictionary(x => x.Key.ToString(), x => x.Value.ToString());
+                if (!_dataSigningService.Check(dictionary, response.Signature))
+                {
+                    Log.Error("Check Signature Failure");
+                    ErrorMessage = "An error has occurred. Payment not processed. If you experience further problems, contact us.";
+                    return RedirectToAction("Index", "Home");
+                }
+                //ViewBag.PaymentDictionary = dictionary; //Debugging. Remove when not needed
+                ProcessPaymentEvent(response, dictionary); //TODO: Do we want to try to write cancel events?
+
+                return View(response);
+            }
+        }
+        
+        [HttpPost]
+        [AllowAnonymous]
+        [IgnoreAntiforgeryToken]
+        public async Task<ActionResult> ProviderNotify(ReceiptResponseModel response)
+        {
+            Log.ForContext("response", response, true).Information("Provider Notification Received");
+
+            // check signature
+            var dictionary = Request.Form.ToDictionary(x => x.Key.ToString(), x => x.Value.ToString());
+            if (!_dataSigningService.Check(dictionary, response.Signature))
+            {
+                Log.Error("Check Signature Failure");
+            }
+            else
+            {
+                ProcessPaymentEvent(response, dictionary);
+
+                //Do payment stuff.
+                var order = _context.Orders.SingleOrDefault(a => a.Id == response.Req_Reference_Number);
+                if (order == null)
+                {
+                    Log.Error("Order not found {0}", response.Req_Reference_Number);
+                }
+                else
+                {
+                    if (response.Decision == ReplyCodes.Accept)
+                    {
+                        if (order.Status == OrderStatusCodes.AwaitingPayment)
+                        {
+                            order.Status = OrderStatusCodes.Paid;
+                            _context.SaveChanges(true);

```

Comment:
@jSylvestre I wonder if you'd want to do order.Payment = payment, and then a single save would save both the payment info as well as the updated order.  That seems simple and atomic and would make future querying easier.

Reply:
@srkirkland The PaymentEvents contains all failures and cancels as well, so I \'m not sure if that would work.

Reply:
@jSylvestre but in this case you only make the order paid if you know you have a successful order, so i don't know why it matters, you'll only have the one success for any given order.